doc/contributing/primordials.md

```diff
@@ -563,6 +563,62 @@ SafePromiseAll(array, someFunction); // Same as the above, but more efficient.
 
 </details>
 
+### Async functions
+
+<details>
+
+<summary>Awaiting or returning a Promise within the body of an async function
+         depends on user-mutable <code>%Promise.prototype%</code>
+         properties</summary>
+
+Within an async function body, the expression `await v` is internally expanded
+into code that looks like the following:
+
+```js
+let promise;
+if (isPromise(v) && v.constructor === Promise) {
+  promise = v;
+} else if (typeof v === 'object' && typeof v.then === 'function') {
+  promise = new Promise((resolve, reject) => {
+    v.then(resolve, reject);
+  });
+} else {
+  promise = Promise.resolve(v);
+}
+AsynchronouslyReturnOrThrowPromiseResult(promise);
+```
+
+By modifying the `constructor` property of `%Promise.prototype%`, the await
+expression can be made to look up and call the `then` property of the awaited
+Promise, which is also user-mutable via `%Promise.prototype%`.
+
+Within an async function body, the statement `return v` is internally expanded
+into code that looks like the following:
+
+```js
+let promise;
+if (typeof v === 'object' && typeof v.then === 'function') {
+  promise = new Promise((resolve, reject) => {
+    v.then(resolve, reject);
+  });
+} else {
+  promise = Promise.resolve(v);
+}
+return promise;
+```
+
+Returning a Promise from an async function body always looks up and calls the
+`then` property of the returned Promise, which is user-mutable via
+`%Promise.prototype%`.
```

Comment:
I think this would be more informative if you shows a code trying to exploit that, having the implementation details is hardly helpful and less convincing IMO